x/token: Total supply is not updated

When exporting the genesis for hard-fork, the following error occurs.
It's because the `x/token` module doesn't update the `supply.Total` when issuing new token. 

```bash
$ panacead export --for-zero-height --height=8217100
panic: invariant broken: supply: total supply invariant
        sum of accounts coins: 3184222347031377umed,1000000000000uyjt344
        supply.Total:          3184222347031377umed
        CRITICAL please submit the following transaction:
                 tx crisis invariant-broken supply total-supply
goroutine 1 [running]:
github.com/cosmos/cosmos-sdk/x/crisis/internal/keeper.Keeper.AssertInvariants(0xc000e42a00, 0xb, 0x10, 0xc000e16070, 0x14fa318, 0xc000f0da00, 0x14fa340, 0xc000f0da50, 0xc000fe37d0, 0x6, ...)
        /home/centos/go/pkg/mod/github.com/cosmos/cosmos-sdk@v0.37.14/x/crisis/internal/keeper/keeper.go:77 +0x4c6
github.com/medibloc/panacea-core/app.(*PanaceaApp).prepForZeroHeightGenesis(0xc00001bb80, 0x1506948, 0xc000134190, 0x1512b10, 0xc0015d4140, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/centos/panacea-core/app/export.go:57 +0x278
github.com/medibloc/panacea-core/app.(*PanaceaApp).ExportAppStateAndValidators(0xc00001bb80, 0x7d6201, 0x1de7410, 0x0, 0x0, 0x0, 0xc000f25900, 0x1, 0x0, 0x0, ...)
        /home/centos/panacea-core/app/export.go:23 +0x348
main.exportAppStateAndTMValidators(0x1507198, 0xc000f08f80, 0x15141b8, 0xc000138318, 0x0, 0x0, 0x7d620c, 0xc0000cb901, 0x1de7410, 0x0, ...)
        /home/centos/panacea-core/cmd/panacead/main.go:99 +0x1a8
github.com/cosmos/cosmos-sdk/server.ExportCmd.func1(0xc000e42280, 0xc000e3a200, 0x0, 0x2, 0x0, 0x0)
        /home/centos/go/pkg/mod/github.com/cosmos/cosmos-sdk@v0.37.14/server/export.go:65 +0x35d
github.com/spf13/cobra.(*Command).execute(0xc000e42280, 0xc000e3a1c0, 0x2, 0x2, 0xc000e42280, 0xc000e3a1c0)
        /home/centos/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:826 +0x472
github.com/spf13/cobra.(*Command).ExecuteC(0xc000177680, 0xc000e39cc0, 0x11e6fad, 0xc0000cbec8)
        /home/centos/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:914 +0x30b
github.com/spf13/cobra.(*Command).Execute(...)
        /home/centos/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:864
github.com/tendermint/tendermint/libs/cli.Executor.Execute(0xc000177680, 0x13c04b0, 0x11cbb1c, 0x10)
        /home/centos/go/pkg/mod/github.com/tendermint/tendermint@v0.32.13/libs/cli/setup.go:89 +0x3c
main.main()
        /home/centos/panacea-core/cmd/panacead/main.go:73 +0x878
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,7 +2,10 @@
 
 ## Unreleased
 
-TBD
+### Bug fixes
+
+- [\#95](https://github.com/medibloc/panacea-core/pull/95) x/token: Fix to update the total supply of all coins when issuing new tokens
+
 
 ## [v1.3.3](https://github.com/medibloc/panacea-core/releases/tag/v1.3.3) - 2020-12-03
 

--- a/app/app.go
+++ b/app/app.go
@@ -222,6 +222,7 @@ func NewPanaceaApp(
 		keys[token.StoreKey],
 		app.cdc,
 		app.bankKeeper,
+		app.supplyKeeper,
 	)
 
 	// register the staking hooks

--- a/x/token/keeper/keeper.go
+++ b/x/token/keeper/keeper.go
@@ -55,7 +55,7 @@ func (k tokenKeeper) SetToken(ctx sdk.Context, symbol types.Symbol, token types.
 	newTotal := supply.GetTotal().Add(newCoins)
 	k.supplyKeeper.SetSupply(ctx, supply.SetTotal(newTotal))
 
-	// Deposit the total supply to the owner account
+	// Deposit the total supply of the new coin to the owner account
 	if _, err := k.bankKeeper.AddCoins(ctx, token.OwnerAddress, newCoins); err != nil {
 		panic(err)
 	}

--- a/x/token/keeper/keeper.go
+++ b/x/token/keeper/keeper.go
@@ -24,15 +24,17 @@ type tokenKeeper struct {
 	cdc *codec.Codec
 
 	// External keepers
-	bankKeeper types.BankKeeper
+	bankKeeper   types.BankKeeper
+	supplyKeeper types.SupplyKeeper
 }
 
 // NewKeeper creates a new instance of the token Keeper
-func NewKeeper(storeKey sdk.StoreKey, cdc *codec.Codec, bankKeeper types.BankKeeper) Keeper {
+func NewKeeper(storeKey sdk.StoreKey, cdc *codec.Codec, bankKeeper types.BankKeeper, supplyKeeper types.SupplyKeeper) Keeper {
 	return tokenKeeper{
-		storeKey:   storeKey,
-		cdc:        cdc,
-		bankKeeper: bankKeeper,
+		storeKey:     storeKey,
+		cdc:          cdc,
+		bankKeeper:   bankKeeper,
+		supplyKeeper: supplyKeeper,
 	}
 }
 
@@ -46,8 +48,15 @@ func (k tokenKeeper) SetToken(ctx sdk.Context, symbol types.Symbol, token types.
 	bz := k.cdc.MustMarshalBinaryLengthPrefixed(token)
 	store.Set(key, bz)
 
+	newCoins := sdk.NewCoins(token.TotalSupply)
+
+	// Update the total supply of all coins
+	supply := k.supplyKeeper.GetSupply(ctx)
+	newTotal := supply.GetTotal().Add(newCoins)
+	k.supplyKeeper.SetSupply(ctx, supply.SetTotal(newTotal))
+
 	// Deposit the total supply to the owner account
-	if _, err := k.bankKeeper.AddCoins(ctx, token.OwnerAddress, sdk.NewCoins(token.TotalSupply)); err != nil {
+	if _, err := k.bankKeeper.AddCoins(ctx, token.OwnerAddress, newCoins); err != nil {
 		panic(err)
 	}
 }

--- a/x/token/types/expected_keepers.go
+++ b/x/token/types/expected_keepers.go
@@ -1,8 +1,17 @@
 package types
 
-import sdk "github.com/cosmos/cosmos-sdk/types"
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/supply/exported"
+)
 
 // BankKeeper defines the expected bank Keeper
 type BankKeeper interface {
 	AddCoins(ctx sdk.Context, addr sdk.AccAddress, amt sdk.Coins) (sdk.Coins, sdk.Error)
+}
+
+// SupplyKeeper defines the expected bank Keeper
+type SupplyKeeper interface {
+	GetSupply(ctx sdk.Context) (supply exported.SupplyI)
+	SetSupply(ctx sdk.Context, supply exported.SupplyI)
 }

